[Bug]: "Observed a panic" with FlexibleServer set "authentication.passwordAuthEnabled: false."

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Affected Resource(s)

* dbforpostgresql.azure.upbound.io/v1beta2 FlexibleServer

### Resource MRs required to reproduce the bug

```
apiVersion: azure.upbound.io/v1beta1
kind: ResourceGroup
metadata:
  name: debug
  labels:
    testing.upbound.io/example-name: debug
spec:
  deletionPolicy: Delete
  forProvider:
    location: "Germany West Central"
---
apiVersion: dbforpostgresql.azure.upbound.io/v1beta2
kind: FlexibleServer
metadata:
  labels:
    testing.upbound.io/example-name: debug
  name: debug
spec:
  deletionPolicy: Delete
  forProvider:
    location: Germany West Central
    authentication:
      activeDirectoryAuthEnabled: true
      passwordAuthEnabled: false
      tenantId: 1234
    resourceGroupNameSelector:
      matchLabels:
        testing.upbound.io/example-name: debug
    skuName: GP_Standard_D2s_v3
    storageMb: 32768
    version: "16"
```

### Steps to Reproduce

Simply apply the provided resources. Also the `provider-azure-dbforpostgresql@v1.1.0` and using the `apiVersion: v1beta1` are affected. 

If at `passwordAuthEnabled` is set to `true` and the required other params are configured (administratorLogin, autoGeneratePassword, administratorPasswordSecretRef etc.) the issue does not occur. 

### What happened?

I'm using v1.2 [upbound/provider-azure-dbforpostgresql@v1.2.0](https://marketplace.upbound.io/providers/upbound/provider-azure-dbforpostgresql/v1.2.0) with the latest crossplane (v1.16.0)

The stacktrace tells us that upjet wants to call the "sensitive" attributes here:
https://github.com/crossplane/upjet/blob/main/pkg/resource/sensitive.go#L70
```
github.com/crossplane/upjet/pkg/resource.GetConnectionDetails(0xc0038d1c30?, {0xd96bb40?, 0xc00599ca00?}, 0xc005052b40)
	github.com/crossplane/upjet@v1.4.0-rc.0.0.20240515193317-92d1af84d242/pkg/resource/sensitive.go:73 +0x76
```

which leads to a call here: https://github.com/crossplane-contrib/provider-upjet-azure/blob/main/config/postgresql/config.go#L95

```
github.com/upbound/provider-azure/config/postgresql.Configure.func9.1(0xc0067acab0?)
	github.com/upbound/provider-azure/config/postgresql/config.go:95 +0x315
```

as with the setting `authentication.passwordAuthEnabled: false` no password is set (as AD / Entra ID auth is configured only), the call leads to accessing an attribute which is not present (therefore the panic). 


### Relevant Error Output Snippet

```shell
2024-06-09T09:56:04Z	DEBUG	provider-azure	Starting	{"sync-interval": "1h0m0s", "poll-interval": "10m0s", "poll-jitter": "30s", "max-reconcile-rate": 100}
2024-06-09T09:56:07Z	INFO	provider-azure	Beta feature enabled	{"flag": "EnableBetaManagementPolicies"}
2024-06-09T09:56:07Z	DEBUG	provider-azure	Calling the inner handler for Create event.	{"gvk": "dbforpostgresql.azure.upbound.io/v1beta1, Kind=FlexibleServer", "name": "debug", "queueLength": 0}
2024-06-09T09:56:08Z	DEBUG	provider-azure	Reconciling	{"controller": "managed/dbforpostgresql.azure.upbound.io/v1beta1, kind=flexibleserver", "request": {"name":"debug"}}
2024-06-09T09:56:08Z	DEBUG	provider-azure	Connecting to the service provider	{"uid": "320a8a70-d850-43ab-af70-3fcf6e28782a", "name": "debug", "gvk": "dbforpostgresql.azure.upbound.io/v1beta1, Kind=FlexibleServer"}
2024-06-09T09:56:12Z	DEBUG	provider-azure	Instance state not found in cache, reconstructing...	{"uid": "320a8a70-d850-43ab-af70-3fcf6e28782a", "name": "debug", "gvk": "dbforpostgresql.azure.upbound.io/v1beta1, Kind=FlexibleServer"}
2024-06-09T09:56:12Z	DEBUG	provider-azure	Observing the external resource	{"uid": "320a8a70-d850-43ab-af70-3fcf6e28782a", "name": "debug", "gvk": "dbforpostgresql.azure.upbound.io/v1beta1, Kind=FlexibleServer"}
E0609 09:56:12.448311       1 runtime.go:79] Observed a panic: &runtime.TypeAssertionError{_interface:(*abi.Type)(0xaf8dac0), concrete:(*abi.Type)(nil), asserted:(*abi.Type)(0xaabba00), missingMethod:""} (interface conversion: interface {} is nil, not string)
goroutine 5142 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0xb2c09e0?, 0xc006b8e8d0})
	k8s.io/apimachinery@v0.29.4/pkg/util/runtime/runtime.go:75 +0x85
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:108 +0xb2
panic({0xb2c09e0?, 0xc006b8e8d0?})
	runtime/panic.go:914 +0x21f
github.com/upbound/provider-azure/config/postgresql.Configure.func9.1(0xc0067acab0?)
	github.com/upbound/provider-azure/config/postgresql/config.go:95 +0x315
github.com/crossplane/upjet/pkg/resource.GetConnectionDetails(0xc0038d1c30?, {0xd96bb40?, 0xc00599ca00?}, 0xc005052b40)
	github.com/crossplane/upjet@v1.4.0-rc.0.0.20240515193317-92d1af84d242/pkg/resource/sensitive.go:73 +0x76
github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKExternal).Observe(0xc00459c0e0, {0xd894c70, 0xc005b109a0}, {0xd9142f0?, 0xc00599ca00?})
	github.com/crossplane/upjet@v1.4.0-rc.0.0.20240515193317-92d1af84d242/pkg/controller/external_tfpluginsdk.go:522 +0x939
github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKAsyncExternal).Observe(0xc006e52780, {0xd894c70, 0xc005b109a0}, {0xd9142f0?, 0xc00599ca00?})
	github.com/crossplane/upjet@v1.4.0-rc.0.0.20240515193317-92d1af84d242/pkg/controller/external_async_tfpluginsdk.go:127 +0x1ea
github.com/crossplane/crossplane-runtime/pkg/reconciler/managed.(*Reconciler).Reconcile(0xc0021aee00, {0xd894bc8, 0xc005fe7d40}, {{{0x0, 0x0}, {0xc0038d1b30, 0x5}}})
	github.com/crossplane/crossplane-runtime@v1.16.0-rc.2.0.20240510094504-3f697876fa57/pkg/reconciler/managed/reconciler.go:914 +0x2abf
github.com/crossplane/crossplane-runtime/pkg/ratelimiter.(*Reconciler).Reconcile(0xc005998730, {0xd894bc8, 0xc005fe7d40}, {{{0x0?, 0x5?}, {0xc0038d1b30?, 0xc003876d08?}}})
	github.com/crossplane/crossplane-runtime@v1.16.0-rc.2.0.20240510094504-3f697876fa57/pkg/ratelimiter/reconciler.go:54 +0x151
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xd8b4538?, {0xd894bc8?, 0xc005fe7d40?}, {{{0x0?, 0xb?}, {0xc0038d1b30?, 0x0?}}})
	sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc005989a40, {0xd894c00, 0xc00041df90}, {0xb8f10c0?, 0xc00339aa60?})
	sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:316 +0x3cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc005989a40, {0xd894c00, 0xc00041df90})
	sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266 +0x1af
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 4282
	sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:223 +0x565
```


### Crossplane Version

1.16.0

### Provider Version

1.2.0

### Kubernetes Version

v1.26.9

### Kubernetes Distribution

OpenShift & AKS

### Additional Info

This issue here https://github.com/crossplane-contrib/provider-upjet-azure/issues/667 is related, as it's also running into a situation where an attribute is not present & a panic. 

After the panic the FlexibelServer MR is not updated anymore (e.g. .status is not updated) - also subsequent changes to the resource cannot be processed anymore. 

I see two potential ways how this issue might be resolved:
1. https://github.com/crossplane/upjet/blob/main/pkg/resource/sensitive.go#L70 <- Either the caller of the surrounding function or here should have awareness when a provider does not have credentials (e.g. when external AuthX provider like Entra ID are used). But as this is provider specific, I'm not sure if this makes sense. 
2. Or https://github.com/crossplane-contrib/provider-upjet-azure/blob/main/config/postgresql/config.go#L90 requires a condition which verifies that `authentication.passwordAuthEnabled` is set to `true` - otherwise either returning `nil` or an empty map. We can see that in other provides such conditions were done: https://github.com/crossplane-contrib/provider-upjet-azure/blame/0a511b8f5330fc8a8dec7dea0c76e2ef4e16ffcc/config/sql/config.go#L39

Maybe @phuhomet or @ulucinar, would it be possible for you to give a bit of insight about when https://github.com/crossplane-contrib/provider-upjet-azure/blame/main/config/postgresql/config.go#L90 is called, if there is a possible workaround and how to best approach when no actual connection details (like with Entra ID) are present at the level of a FlexibleServer?

Hi @niiku,

Thank you for bringing up this, I can reproduce the bug for the `v1beta1` version of the resource with the information you provided.
```
panic: interface conversion: interface {} is nil, not string

goroutine 189957 [running]:
github.com/upbound/provider-azure/config/postgresql.Configure.func9.1(0x14007767c50?)
	github.com/upbound/provider-azure/config/postgresql/config.go:95 +0x30c
github.com/crossplane/upjet/pkg/resource.GetConnectionDetails(0x1401cef69a0?, {0x10f928ba0?, 0x1401be9e500?}, 0x14003b726c0)
	github.com/crossplane/upjet@v1.4.1-0.20240530140005-40733472e6d2/pkg/resource/sensitive.go:70 +0x60
github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKExternal).Create(0x1401cef69a0, {0x10f811420, 0x14002245030}, {0x10f8d1680?, 0x1401be9e500})
	github.com/crossplane/upjet@v1.4.1-0.20240530140005-40733472e6d2/pkg/controller/external_tfpluginsdk.go:642 +0x1f0
github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKAsyncExternal).Create.func1()
	github.com/crossplane/upjet@v1.4.1-0.20240530140005-40733472e6d2/pkg/controller/external_async_tfpluginsdk.go:149 +0x120
created by github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKAsyncExternal).Create in goroutine 34581
	github.com/crossplane/upjet@v1.4.1-0.20240530140005-40733472e6d2/pkg/controller/external_async_tfpluginsdk.go:145 +0x14c
```

--- a/config/postgresql/config.go
+++ b/config/postgresql/config.go
@@ -88,12 +88,19 @@ func Configure(p *config.Provider) {
 			IgnoredFields: []string{"ssl_enforcement", "storage_profile"},
 		}
 		r.Sensitive.AdditionalConnectionDetailsFn = func(attr map[string]interface{}) (map[string][]byte, error) {
-			return map[string][]byte{
-				xpv1.ResourceCredentialsSecretUserKey:     []byte(fmt.Sprintf("%s@%s", attr["administrator_login"], attr["name"])),
+			conn := map[string][]byte{
 				xpv1.ResourceCredentialsSecretEndpointKey: []byte(attr["fqdn"].(string)),
 				xpv1.ResourceCredentialsSecretPortKey:     []byte(strconv.Itoa(postgresqlServerPort)),
-				xpv1.ResourceCredentialsSecretPasswordKey: []byte(attr["administrator_password"].(string)),
-			}, nil
+			}
+			// administrator_login is optional and may not be available depending on the auth type.
+			if _, ok := attr["administrator_login"].(string); ok {
+				conn[xpv1.ResourceCredentialsSecretUserKey] = []byte(fmt.Sprintf("%s@%s", attr["administrator_login"].(string), attr["name"].(string)))
+			}
+			// administrator_password is optional and may not be available depending on the auth type.
+			if _, ok := attr["administrator_password"].(string); ok {
+				conn[xpv1.ResourceCredentialsSecretPasswordKey] = []byte(attr["administrator_password"].(string))
+			}
+			return conn, nil
 		}
 		desc, _ := comments.New("If true, the password will be auto-generated and"+
 			" stored in the Secret referenced by the administratorPasswordSecretRef field.",

--- a/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
+++ b/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
@@ -31,8 +31,7 @@ spec:
     skuName: GP_Standard_D4s_v3
     storageMb: 32768
     storageTier: P30
-    version: "12"
-    zone: "1"
+    version: "16 "
 
 ---
 

--- a/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
+++ b/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
@@ -13,6 +13,7 @@ metadata:
 spec:
   forProvider:
     administratorLogin: psqladmin
+    autoGeneratePassword: true
     administratorPasswordSecretRef:
       key: example-key
       name: example-secret

--- a/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
+++ b/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
@@ -31,7 +31,7 @@ spec:
     skuName: GP_Standard_D4s_v3
     storageMb: 32768
     storageTier: P30
-    version: "16 "
+    version: "16"
 
 ---
 

--- a/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
+++ b/examples/dbforpostgresql/v1beta2/flexibleserver.yaml
@@ -40,6 +40,7 @@ kind: PrivateDNSZone
 metadata:
   annotations:
     meta.upbound.io/example-id: dbforpostgresql/v1beta2/flexibleserver
+    crossplane.io/external-name: example-upbound.postgres.database.azure.com
   labels:
     testing.upbound.io/example-name: example
   name: example

